### PR TITLE
Ledger refactoring

### PR DESCRIPTION
**Pull request by Gilthoniel**
This adds a _core_ module where all the ledger related packages will be moved eventually. For this PR, it only implements very simple dummy packages for the interfaces, then the already implemented ones will be moved progressively.

### CODE REVIEW

File: core/ordering/mod.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+// to order transactions from a pool.
+type Service interface {
+	// Listen opens the endpoints of the service so that other participants can
+	// contact the node.
+	Listen() error
+
+	// Close closes the service.
+	Close() error
+
+	// GetProof must return a proof of the value of the provided key.
```
Comment (nkcr) [inline core/ordering/mod.go:37]: ...value AT the...

File: core/tap/pool/mem/mod_test.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+func TestPool_Add(t *testing.T) {
+	pool := NewPool()
+
+	err := pool.Add(fakeTx{id: []byte{1}})
+	require.NoError(t, err)
+
+	err = pool.Add(fakeTx{id: []byte{2}})
+	require.NoError(t, err)
+
+	// A transaction that exists in the active queue can simply be overriden
```
Comment (nkcr) [inline core/tap/pool/mem/mod_test.go:34]:
```suggestion
	// A transaction that exists in the active queue can simply be overwritten
```

File: core/execution/baremetal/mod.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+func NewExecution(exec execution.Service) BareMetal {
+	return BareMetal{
+		exec: exec,
+	}
+}
+
+// Execute implements execution.Service. It uses the executor to process the
+// incoming transaction and return the result.
+func (bm BareMetal) Execute(tx tap.Transaction, snap store.Snapshot) (execution.Result, error) {
+	res, err := bm.exec.Execute(tx, snap)
```
Comment (nkcr) [inline core/execution/baremetal/mod.go:31]: an execution.Service that uses another execution.Service ? Looks like the snake that bites its own tail.

File: core/tap/anon/mod.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	return nil
+}
+
+// GetArg implements tap.Transaction. It returns the value of the argument if it
+// is set, otherwise nil.
+func (t Transaction) GetArg(key string) []byte {
+	return t.args[key]
+}
+
+// Fingerprint implements serde.Fingerprinter. It writes a deteministic binary
```
Comment (nkcr) [inline core/tap/anon/mod.go:87]:
```suggestion
// Fingerprint implements serde.Fingerprinter. It writes a deterministic binary
```

File: core/ordering/mod.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+// Service is the interface of an ordering service. It provides the primitives
+// to order transactions from a pool.
+type Service interface {
+	// Listen opens the endpoints of the service so that other participants can
+	// contact the node.
+	Listen() error
+
+	// Close closes the service.
+	Close() error
```
[the remark this answers is not in the document]
Comment (Gilthoniel) [reply core/ordering/mod.go:35]: Not sure if that will be useful but right you can do it yes, so I'll rename.

File: core/execution/baremetal/mod.go (hunk 3 above)
Comment (Gilthoniel) [reply core/execution/baremetal/mod.go:31]: Yeah it looks dump right now, but the goal was to register different services (darc, roster, etc...) that are the bare-metal smart contracts and this service takes care of looking up the right one.

Comment (Gilthoniel) [reply core/execution/baremetal/mod.go:31]: Which will come while migrating the other packages.

**Pull request merged**